projects/ardin_mb/visual_navigation_perfect_memory_window.h

```diff
@@ -0,0 +1,109 @@
+#pragma once
+
+// Ardin MB includes
+#include "visual_navigation_base.h"
+
+// BoB Robotics includes
+#include "plog/Log.h"
+#include "navigation/perfect_memory.h"
+#include "navigation/perfect_memory_window.h"
+
+// Standard C++ includes
+#include <algorithm>
+#include <memory>
+
+//----------------------------------------------------------------------------
+// VisualNavigationPerfectMemoryWindow
+//----------------------------------------------------------------------------
+//! Very thin wrapper to allow VisualNavigationBase wrap BoB
+//! navigation algorithms which can be used without additional state etc
+template<typename S = BoBRobotics::Navigation::PerfectMemoryStore::RawImage<>>
+class VisualNavigationPerfectMemoryWindow : public VisualNavigationBase
+{
+    template<typename T>
+    using PerfectMemory = BoBRobotics::Navigation::PerfectMemory<T>;
+
+    typedef std::unique_ptr<BoBRobotics::Navigation::PerfectMemoryWindow::Base> WindowPtr;
+
+public:
+    template<class... Ts>
+    VisualNavigationPerfectMemoryWindow(WindowPtr window, Ts &&... args)
```

Comment:
I might be missing something, but does ``window`` need to be a pointer here?

Reply:
Good point. There's no real need for this class to own a window, it could just take a reference. The ideal solution would be to have this class templated on window type but then you couldn't perfect forward it's constructor parameters as well as the ``PerfectMemory`` ones

Reply:
I'm not saying this is a *good* idea, but I'm wondering whether you could have two sets of perfectly forwarded params if each set was a tuple, i.e.:
```c++
    template<class... Ts, class... Us>
    VisualNavigationPerfectMemoryWindow(WindowPtr window, std::tuple<Ts &&>... args1, std::tuple<Us &&>... args2)
```

Reply:
after hitting nasty scoping issues when trying to use a reference I went the whole hog and switched this to an API styled after the ``std::piecewise_construct`` variant of ``std::map::emplace``:
```c++
using PMWindow = Navigation::PerfectMemoryWindow::DynamicBestMatchGradient;

memory = std::make_unique<VisualNavigationPerfectMemoryWindow<PMWindow>>(
    std::forward_as_tuple(cv::Size(36, 10)),
    std::forward_as_tuple(10, PMWindow::WindowConfig{5, 5, 10, 200}));
```
where the ``VisualNavigationPerfectMemoryWindow`` owns the memory and window objects. In C++14, some horror is required, but I'm pretty happy with the result. As a bonus you can now also select what algorithm to use at runtime from the CLI.

Reply:
That's really cool! It would seem to make sense to have ``VisualNavigationPerfectMemoryWindow`` own everything...